File: usb-bench-c/src/w_iso.c
/*
 * w_iso.c — W-iso benchmark: isochronous UVC streaming on a Logitech Brio.
 *
 * Activates the UVC video-streaming interface (alt-setting 1), then issues
 * a stream of isochronous IN transfers and records the per-transfer RTT and
 * cumulative throughput.
 *
 * Thesis conditions served by this single source file:
 *   C1  native-libusb   compiled for host with system libusb
 *   C2  wasi-libusb     compiled for wasm32-wasip2 with libusb-wasi.a
 *
 * Usage:
 *   w_iso [output.csv] [VID:PID] [iterations]
 *         [--iface N] [--alt N] [--ep 0xNN] [--pkt-size N] [--num-pkts N]
 *         [--condition <name>]
 *
 * Defaults:
 *   output.csv   bench_w_iso.csv
 *   VID:PID      046d:086c  (Logitech Brio 4K)
 *   iterations   200
 *   iface        1
 *   alt          1
 *   ep           0x81
 *   pkt-size     1024
 *   num-pkts     32
 *   condition    BENCH_CONDITION macro ("native-libusb" or "wasi-libusb")
 *
 * Isochronous transfers in libusb are asynchronous.  Each iteration:
 *   1. Allocate a transfer with num_pkts ISO packets.
 *   2. Fill and submit it.
 *   3. Run the event loop until the completion callback fires.
 *   4. Sum the actual_length fields across packets.
 *   5. Record RTT, bytes, and resource usage.
 */

#include "csv.h"

#include <libusb.h>

#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* ── Compile-time defaults ───────────────────────────────────────────────── */
#ifndef BENCH_CONDITION
#  define BENCH_CONDITION "native-libusb"
#endif

#define DEFAULT_OUTPUT     "bench_w_iso.csv"
#define DEFAULT_VID        0x046du   /* Logitech */
#define DEFAULT_PID        0x086cu   /* Brio 4K  */
#define DEFAULT_ITERATIONS 200u
#define DEFAULT_IFACE      1u
#define DEFAULT_ALT        1u
#define DEFAULT_EP         0x81u
#define DEFAULT_PKT_SIZE   1024u
#define DEFAULT_NUM_PKTS   32u

/* ── ISO completion callback ────────────────────────────────────────────── */
/*
 * The callback is called from libusb_handle_events_completed().
 * We store a pointer to a volatile int "completed" flag in user_data.
 * Setting it to 1 makes the event loop in main() exit.
 */
static void LIBUSB_CALL iso_callback(struct libusb_transfer *xfer)
{
    volatile int *completed = (volatile int *)xfer->user_data;
    *completed = 1;
}

/* ── Helpers ─────────────────────────────────────────────────────────────── */

static void die(const char *msg)
{
    fprintf(stderr, "fatal: %s\n", msg);
    exit(1);
}

static void parse_vid_pid(const char *s, uint16_t *vid, uint16_t *pid)
{
    unsigned v = 0, p = 0;
    if (sscanf(s, "%x:%x", &v, &p) != 2)
        die("expected VID:PID in hex, e.g. 046d:086c");
    *vid = (uint16_t)v;
    *pid = (uint16_t)p;
}

static int parse_uint_arg(const char *s)
{
    char *end = NULL;
    long v = strtol(s, &end, 0); /* base 0: accepts 0x prefix */
    return (end != s && v >= 0) ? (int)v : -1;
}

/* ── Main ─────────────────────────────────────────────────────────────────── */

int main(int argc, char **argv)
{
    const char *output     = (argc > 1) ? argv[1] : DEFAULT_OUTPUT;
    const char *vid_pid    = (argc > 2) ? argv[2] : "046d:086c";
    uint32_t    iterations = (argc > 3) ? (uint32_t)atoi(argv[3]) : DEFAULT_ITERATIONS;
    const char *condition  = BENCH_CONDITION;

    uint8_t  iface    = DEFAULT_IFACE;
    uint8_t  alt      = DEFAULT_ALT;
    uint8_t  ep       = DEFAULT_EP;
    uint32_t pkt_size = DEFAULT_PKT_SIZE;
    uint32_t num_pkts = DEFAULT_NUM_PKTS;

    for (int i = 1; i < argc; i++) {
        if (i + 1 < argc) {
            int v = parse_uint_arg(argv[i + 1]);
            if      (strcmp(argv[i], "--iface")     == 0 && v >= 0) iface    = (uint8_t)v;
            else if (strcmp(argv[i], "--alt")       == 0 && v >= 0) alt      = (uint8_t)v;
            else if (strcmp(argv[i], "--ep")        == 0 && v >= 0) ep       = (uint8_t)v;
            else if (strcmp(argv[i], "--pkt-size")  == 0 && v >= 0) pkt_size = (uint32_t)v;
            else if (strcmp(argv[i], "--num-pkts")  == 0 && v >= 0) num_pkts = (uint32_t)v;
            else if (strcmp(argv[i], "--condition") == 0) condition = argv[i + 1];
        }
    }

    uint16_t vid, pid;
    parse_vid_pid(vid_pid, &vid, &pid);

    uint32_t buf_size = pkt_size * num_pkts;

    /* ── libusb init ─────────────────────────────────────────────────────── */
    libusb_context *ctx = NULL;
    if (libusb_init(&ctx) != 0)
        die("libusb_init failed");

    libusb_device_handle *handle =
        libusb_open_device_with_vid_pid(ctx, vid, pid);
    if (!handle) {
        fprintf(stderr, "fatal: device %04x:%04x not found\n", vid, pid);
        libusb_exit(ctx);
        return 1;
    }

    /* Detach kernel UVC driver if present (Linux) */
    if (libusb_kernel_driver_active(handle, iface) == 1)
        libusb_detach_kernel_driver(handle, iface);

    /* Claim the video-streaming interface at alt-setting 0 first */
    int rc = libusb_claim_interface(handle, iface);
    if (rc != 0) {
        fprintf(stderr, "fatal: libusb_claim_interface(%u): %s\n",
                iface, libusb_error_name(rc));
        libusb_close(handle);
        libusb_exit(ctx);
        return 1;
    }

    /* Switch to the alt-setting that has ISO endpoints (alt >= 1) */
    rc = libusb_set_interface_alt_setting(handle, iface, alt);
    if (rc != 0) {
        fprintf(stderr, "fatal: libusb_set_interface_alt_setting(%u,%u): %s\n",
                iface, alt, libusb_error_name(rc));
        libusb_release_interface(handle, iface);
        libusb_close(handle);
        libusb_exit(ctx);
        return 1;
    }

    /* ── CSV logger ─────────────────────────────────────────────────────── */
    FILE *csv = bench_csv_open(output);
    if (!csv)
        die("bench_csv_open failed");

    fprintf(stderr,
            "▶ %04x:%04x  workload=iso  iface=%u  alt=%u  ep=0x%02x  "
            "pkt=%u B  num_pkts=%u  buf=%u B\n",
            vid, pid, iface, alt, ep, pkt_size, num_pkts, buf_size);
    fprintf(stderr,
            "  iterations=%u  condition=%s  → %s\n",
            iterations, condition, output);

    /* ── Allocate ISO transfer and data buffer ───────────────────────────── */
    uint8_t *buf = (uint8_t *)malloc(buf_size);
    if (!buf)
        die("malloc failed");

    struct libusb_transfer *xfer =
        libusb_alloc_transfer((int)num_pkts);
    if (!xfer)
        die("libusb_alloc_transfer failed");

    /* ── Benchmark loop ──────────────────────────────────────────────────── */
    uint64_t total_bytes = 0;

    for (uint32_t iter = 0; iter < iterations; ++iter) {
        bench_rusage_t ru_before, ru_after;
        bench_rusage_now(&ru_before);

        /* Arm the transfer */
        volatile int completed = 0;
        libusb_fill_iso_transfer(
            xfer,
            handle,
            ep,
            buf,
            (int)buf_size,
            (int)num_pkts,
            iso_callback,
            (void *)&completed,
            5000 /* timeout ms */
        );
        libusb_set_iso_packet_lengths(xfer, pkt_size);

        /* ── timed section ─────────────────────────────────────────────── */
        uint64_t t0 = bench_now_ns();

        rc = libusb_submit_transfer(xfer);
        if (rc != 0) {
            fprintf(stderr,
                    "  [%4u/%u] libusb_submit_transfer: %s\n",
                    iter, iterations, libusb_error_name(rc));
            continue;
        }

        /* Pump the event loop until the callback sets completed = 1 */
        while (!completed) {
            rc = libusb_handle_events_completed(ctx, (int *)&completed);
            if (rc != 0 && rc != LIBUSB_ERROR_INTERRUPTED) {
                fprintf(stderr,
                        "  [%4u/%u] libusb_handle_events_completed: %s\n",
                        iter, iterations, libusb_error_name(rc));
                break;
            }
        }

        uint64_t duration_ns = bench_now_ns() - t0;
        /* ── end timed section ─────────────────────────────────────────── */

        bench_rusage_now(&ru_after);

        /* Check transfer status */
        if (xfer->status != LIBUSB_TRANSFER_COMPLETED &&
            xfer->status != LIBUSB_TRANSFER_TIMED_OUT) {
            fprintf(stderr,
                    "  [%4u/%u] transfer status: %d\n",
                    iter, iterations, (int)xfer->status);
            continue;
        }

        /* Sum actual bytes received across all ISO packets */
        uint64_t bytes_received = 0;
        for (uint32_t p = 0; p < num_pkts; ++p) {
            struct libusb_iso_packet_descriptor *pkt =
                &xfer->iso_packet_desc[p];
            if (pkt->status == LIBUSB_TRANSFER_COMPLETED)
                bytes_received += pkt->actual_length;
        }
        total_bytes += bytes_received;

        bench_row_t row = {0};
        row.condition    = condition;
        row.workload     = "iso";
        row.iteration    = iter;
        row.bytes        = bytes_received;
        row.duration_ns  = duration_ns;
        row.user_cpu_us  = ru_after.user_us  - ru_before.user_us;
        row.sys_cpu_us   = ru_after.sys_us   - ru_before.sys_us;
        row.rss_peak_kb  = ru_after.rss_peak_kb;
        row.guest_mem_bytes = bench_guest_mem_bytes();
        row.has_guest_mem   = 0;
#ifdef __wasm__
        row.has_guest_mem   = 1;
#endif
        bench_csv_write(csv, &row);

        if (iter % 20 == 0 || iter == iterations - 1) {
            double mb_s = (bytes_received / 1048576.0) / (duration_ns / 1e9);
            fprintf(stderr,
                    "  [%4u/%u]  bytes=%" PRIu64 "  RTT=%llu µs  "
                    "tput=%.1f MB/s\n",
                    iter, iterations,
                    bytes_received,
                    (unsigned long long)(duration_ns / 1000u),
                    mb_s);
        }
    }

    /* ── Summary ─────────────────────────────────────────────────────────── */
    fprintf(stderr,
            "✓ Done  iterations=%u  total=%.1f MiB  → %s\n",
            iterations, total_bytes / 1048576.0, output);

    /* ── Cleanup ─────────────────────────────────────────────────────────── */
    libusb_free_transfer(xfer);
    free(buf);

    /* Switch back to alt-setting 0 (releases ISO bandwidth) */
    libusb_set_interface_alt_setting(handle, iface, 0);
    libusb_release_interface(handle, iface);
    libusb_close(handle);
    libusb_exit(ctx);
    bench_csv_close(csv);

    return 0;
}

File: usb-bench-rs/src/bin/w_iso.rs
//! W-iso benchmark — isochronous throughput on a Logitech Brio webcam.
//!
//! Submits isochronous IN transfers on the UVC Video Streaming interface and
//! records sustained throughput (bytes/s) and per-transfer latency.
//!
//! Native path uses `libusb1-sys` directly because `rusb 0.9` does not expose
//! a synchronous isochronous API; the raw ISO async submit+await loop is
//! implemented here behind the `IsoDevice` abstraction so that `run_benchmark`
//! is 100% shared between native and WASM.
//!
//! # Conditions served by this binary
//! | Build target        | Condition | Default `--condition` |
//! |---------------------|-----------|-----------------------|
//! | native host         | C3        | `native-rusb`         |
//! | wasm32-wasip2       | C5        | `wasi-raw-wit`        |
//!
//! Default device: Logitech Brio `046d:086c` (VS interface 1, alt 1,
//! endpoint 0x81, packet stride 1024 B, 32 packets per transfer ≈ 32 KiB).
//!
//! # Usage
//! ```text
//! w_iso [output.csv] [VID:PID] [iterations] [--condition <name>]
//!       [--iface N] [--alt N] [--ep 0xNN] [--pkt-size N] [--num-pkts N]
//!
//! Defaults:
//!   output.csv       bench_w_iso.csv
//!   VID:PID          046d:086c  (Logitech Brio)
//!   iterations       200
//!   condition        native-rusb  (native) | wasi-raw-wit  (wasm)
//!   --iface          1
//!   --alt            1
//!   --ep             0x81
//!   --pkt-size       1024
//!   --num-pkts       32
//! ```

use std::path::Path;
use std::time::Instant;

use anyhow::{anyhow, Result};

use usb_bench_rs::csv::{self, Logger, Row, Rusage};

// ---------------------------------------------------------------------------
// Transport selection — the ONLY cfg-gated section in this file
// ---------------------------------------------------------------------------

// C3 (native) and C4 (wasm32-wasip2 + c4-rusb-wasi) → rusb-based native impl
#[cfg(any(not(target_family = "wasm"), feature = "c4-rusb-wasi"))]
use native::IsoDevice as ActiveDevice;

// C5 (wasm32-wasip2, raw WIT) → WIT-based wasm impl
#[cfg(all(target_family = "wasm", not(feature = "c4-rusb-wasi")))]
use wasm::IsoDevice as ActiveDevice;

// ---------------------------------------------------------------------------
// Native transport — libusb1-sys ISO async submit+await — C3 and C4
// ---------------------------------------------------------------------------

#[cfg(any(not(target_family = "wasm"), feature = "c4-rusb-wasi"))]
mod native {
    use anyhow::{anyhow, Result};
    use libusb1_sys::constants::*;
    use libusb1_sys::*;
    use std::os::raw::{c_int, c_void};

    // ── Callback set by libusb when the transfer completes ──────────────────
    extern "system" fn iso_cb(xfer: *mut libusb_transfer) {
        // `user_data` points to the `completed` c_int on the caller's stack.
        // The callback runs synchronously inside `libusb_handle_events_*`,
        // which is called from the same thread, so there is no data race.
        unsafe {
            let completed = (*xfer).user_data as *mut c_int;
            *completed = 1;
        }
    }

    // ── Helper: turn a libusb error code into an anyhow::Error ──────────────
    fn check(rc: c_int, ctx: &str) -> Result<()> {
        if rc == 0 {
            Ok(())
        } else {
            Err(anyhow!("{ctx}: libusb error {rc}"))
        }
    }

    pub struct IsoDevice {
        ctx: *mut libusb_context,
        handle: *mut libusb_device_handle,
        iface: c_int,
        ep_in: u8,
        pub packet_size: u32,
        pub num_packets: u32,
    }

    impl IsoDevice {
        pub fn open(
            vid: u16,
            pid: u16,
            iface: u8,
            alt: u8,
            ep_in: u8,
            packet_size: u32,
            num_packets: u32,
        ) -> Result<Self> {
            unsafe {
                let mut ctx: *mut libusb_context = std::ptr::null_mut();
                check(libusb_init(&mut ctx), "libusb_init")?;

                let handle = libusb_open_device_with_vid_pid(ctx, vid, pid);
                if handle.is_null() {
                    libusb_exit(ctx);
                    return Err(anyhow!("device {:04x}:{:04x} not found", vid, pid));
                }

                // Detach kernel driver if active (video4linux / uvcvideo on Linux)
                if libusb_kernel_driver_active(handle, iface as c_int) == 1 {
                    let _ = libusb_detach_kernel_driver(handle, iface as c_int);
                }

                check(libusb_set_configuration(handle, 1), "set_configuration")?;
                check(libusb_claim_interface(handle, iface as c_int), "claim_interface")?;
                check(
                    libusb_set_interface_alt_setting(handle, iface as c_int, alt as c_int),
                    "set_interface_alt_setting",
                )?;

                Ok(IsoDevice {
                    ctx,
                    handle,
                    iface: iface as c_int,
                    ep_in,
                    packet_size,
                    num_packets,
                })
            }
        }

        /// Submit one ISO IN transfer and wait for it.
        /// Returns the total number of bytes actually transferred.
        pub fn capture(&self) -> Result<u64> {
            let total = (self.packet_size * self.num_packets) as usize;
            let mut buf = vec![0u8; total];
            let mut completed: c_int = 0;

            unsafe {
                let xfer = libusb_alloc_transfer(self.num_packets as c_int);
                if xfer.is_null() {
                    return Err(anyhow!("libusb_alloc_transfer failed (OOM)"));
                }

                libusb_fill_iso_transfer(
                    xfer,
                    self.handle,
                    self.ep_in,
                    buf.as_mut_ptr(),
                    total as c_int,
                    self.num_packets as c_int,
                    iso_cb,
                    &mut completed as *mut c_int as *mut c_void,
                    5_000, // timeout ms
                );
                libusb_set_iso_packet_lengths(xfer, self.packet_size);

                let rc = libusb_submit_transfer(xfer);
                if rc != 0 {
                    libusb_free_transfer(xfer);
                    return Err(anyhow!("libusb_submit_transfer: {rc}"));
                }

                // Pump the event loop until the callback sets `completed = 1`.
                while completed == 0 {
                    libusb_handle_events_completed(self.ctx, &mut completed);
                }

                // Transfer status check
                let status = (*xfer).status;
                let actual = (*xfer).actual_length as u64;
                libusb_free_transfer(xfer);

                if status != LIBUSB_TRANSFER_COMPLETED as c_int
                    && status != LIBUSB_TRANSFER_TIMED_OUT as c_int
                {
                    return Err(anyhow!("iso transfer failed with status {status}"));
                }

                Ok(actual)
            }
        }
    }

    impl Drop for IsoDevice {
        fn drop(&mut self) {
            unsafe {
                let _ = libusb_release_interface(self.handle, self.iface);
                libusb_close(self.handle);
                libusb_exit(self.ctx);
            }
        }
    }

    // Send + Sync are needed so `run_benchmark` can be called generically.
    // Safety: IsoDevice is used from a single thread only.
    unsafe impl Send for IsoDevice {}
    unsafe impl Sync for IsoDevice {}
}

// ---------------------------------------------------------------------------
// WASM transport (raw WIT bindings) — C5 only
// ---------------------------------------------------------------------------

#[cfg(all(target_family = "wasm", not(feature = "c4-rusb-wasi")))]
mod wasm {
    use anyhow::{anyhow, Result};

    // WIT types from the central bindings module (one generate! per binary).
    use usb_bench_rs::bindings::component::usb::configuration::ConfigValue;
    use usb_bench_rs::bindings::component::usb::device::{list_devices, DeviceHandle};
    use usb_bench_rs::bindings::component::usb::errors::LibusbError;
    use usb_bench_rs::bindings::component::usb::transfers::{
        await_transfer, TransferOptions, TransferSetup, TransferType,
    };

    pub struct IsoDevice {
        handle: DeviceHandle,
        iface: u8,
        ep_in: u8,
        pub packet_size: u32,
        pub num_packets: u32,
    }

    impl IsoDevice {
        pub fn open(
            vid: u16,
            pid: u16,
            iface: u8,
            alt: u8,
            ep_in: u8,
            packet_size: u32,
            num_packets: u32,
        ) -> Result<Self> {
            let devices =
                list_devices().map_err(|e| anyhow!("list_devices: {:?}", e))?;
            let (dev, _, _) = devices
                .into_iter()
                .find(|(_, d, _)| d.vendor_id == vid && d.product_id == pid)
                .ok_or_else(|| anyhow!("device {:04x}:{:04x} not found", vid, pid))?;
            let handle = dev.open().map_err(|e| anyhow!("open: {:?}", e))?;
            handle
                .set_configuration(ConfigValue::Value(1))
                .map_err(|e| anyhow!("set_configuration: {:?}", e))?;
            if handle.kernel_driver_active(iface).unwrap_or(false) {
                let _ = handle.detach_kernel_driver(iface);
            }
            handle
                .claim_interface(iface)
                .map_err(|e| anyhow!("claim_interface: {:?}", e))?;
            handle
                .set_interface_altsetting(iface, alt)
                .map_err(|e| anyhow!("set_interface_altsetting: {:?}", e))?;
            Ok(IsoDevice {
                handle,
                iface,
                ep_in,
                packet_size,
                num_packets,
            })
        }

        pub fn capture(&self) -> Result<u64> {
            let total = self.packet_size * self.num_packets;
            let setup = TransferSetup {
                bm_request_type: 0,
                b_request: 0,
                w_value: 0,
                w_index: 0,
            };
            let opts = TransferOptions {
                endpoint: self.ep_in,
                timeout_ms: 5_000,
                stream_id: 0,
                iso_packets: self.num_packets,
            };
            let xfer = self
                .handle
                .new_transfer(TransferType::Isochronous, setup, total, opts)
                .map_err(|e| anyhow!("new_transfer: {:?}", e))?;
            xfer.submit_transfer(&[])
                .map_err(|e| anyhow!("submit_transfer: {:?}", e))?;
            // Timeout is not fatal for ISO — the macOS UVC driver holds the
            // interface and all transfers time out with 0 bytes.  Accept it
            // the same way the native path accepts LIBUSB_TRANSFER_TIMED_OUT.
            let result = match await_transfer(&xfer) {
                Ok(r) => r,
                Err(LibusbError::Timeout) => return Ok(0),
                Err(e) => return Err(anyhow!("await_transfer: {:?}", e)),
            };
            // Sum actual lengths of individual ISO packets
            let actual: u64 = result
                .packets
                .iter()
                .map(|p| p.actual_length as u64)
                .sum();
            Ok(actual)
        }
    }

    impl Drop for IsoDevice {
        fn drop(&mut self) {
            let _ = self.handle.release_interface(self.iface);
        }
    }
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

fn main() -> Result<()> {
    let cfg = Config::from_args(std::env::args().collect())?;
    let mut logger = Logger::open(Path::new(&cfg.output))?;
    let device = ActiveDevice::open(
        cfg.vid,
        cfg.pid,
        cfg.iface,
        cfg.alt,
        cfg.ep_in,
        cfg.packet_size,
        cfg.num_packets,
    )?;
    run_benchmark(device, &cfg, &mut logger)
}

// ---------------------------------------------------------------------------
// Shared benchmark loop (identical for native and wasm)
// ---------------------------------------------------------------------------

fn run_benchmark(
    device: ActiveDevice,
    cfg: &Config,
    logger: &mut Logger,
) -> Result<()> {
    let bytes_per_iter = device.packet_size as u64 * device.num_packets as u64;
    eprintln!(
        "▶ {vid:04x}:{pid:04x}  workload=iso  iface={iface} alt={alt} \
         ep=0x{ep:02x}  pkt={pkt}×{npkt}={kib} KiB",
        vid = cfg.vid,
        pid = cfg.pid,
        iface = cfg.iface,
        alt = cfg.alt,
        ep = cfg.ep_in,
        pkt = device.packet_size,
        npkt = device.num_packets,
        kib = bytes_per_iter / 1024,
    );
    eprintln!(
        "  condition={cond}  iterations={it}  → {out}",
        cond = cfg.condition,
        it = cfg.iterations,
        out = cfg.output,
    );

    let mut total_bytes: u64 = 0;
    let loop_start = Instant::now();

    for iter in 0..cfg.iterations {
        let ru_before = Rusage::snapshot();

        // ── timed section ────────────────────────────────────────────────────
        let t0 = Instant::now();
        let actual_bytes = device.capture()?;
        let duration_ns = t0.elapsed().as_nanos() as u64;
        // ── end timed section ────────────────────────────────────────────────

        let ru_after = Rusage::snapshot();
        let delta = ru_after.delta(&ru_before);

        total_bytes += actual_bytes;

        let mut row = Row::new(cfg.condition.as_str(), "iso");
        row.iteration = iter;
        row.bytes = actual_bytes;
        row.duration_ns = duration_ns;
        row.user_cpu_us = delta.user_us;
        row.sys_cpu_us = delta.sys_us;
        row.rss_peak_kb = delta.rss_peak_kb;
        row.guest_mem_bytes = csv::guest_mem_bytes();
        logger.write(&row)?;

        if iter % 20 == 0 || iter == cfg.iterations - 1 {
            let mb_s = (actual_bytes as f64 / 1_048_576.0)
                / (duration_ns as f64 / 1_000_000_000.0);
            let rtt_us = duration_ns / 1_000;
            eprintln!(
                "  [{iter:3}/{it}]  {mb_s:6.1} MB/s  RTT={rtt_us} µs  bytes={actual_bytes}",
                it = cfg.iterations,
            );
        }
    }

    let wall_secs = loop_start.elapsed().as_secs_f64();
    let avg_mb_s = (total_bytes as f64 / 1_048_576.0) / wall_secs;
    eprintln!(
        "✓ Done  total={total:.1} MiB  avg={avg:.1} MB/s  wall={wall:.2} s  → {out}",
        total = total_bytes as f64 / 1_048_576.0,
        avg = avg_mb_s,
        wall = wall_secs,
        out = cfg.output,
    );
    Ok(())
}

// ---------------------------------------------------------------------------
// CLI configuration
// ---------------------------------------------------------------------------

struct Config {
    output: String,
    vid: u16,
    pid: u16,
    iterations: u32,
    condition: String,
    iface: u8,
    alt: u8,
    ep_in: u8,
    packet_size: u32,
    num_packets: u32,
}

impl Config {
    fn from_args(args: Vec<String>) -> Result<Self> {
        let output = args
            .get(1)
            .map(String::as_str)
            .unwrap_or("bench_w_iso.csv")
            .to_owned();
        let vid_pid = args.get(2).map(String::as_str).unwrap_or("046d:086c");
        let iterations: u32 = args.get(3).and_then(|s| s.parse().ok()).unwrap_or(200);

        let condition = named_arg(&args, "--condition")
            .unwrap_or_else(|| default_condition().to_owned());

        let iface: u8 = named_arg(&args, "--iface")
            .and_then(|s| s.parse().ok())
            .unwrap_or(1);
        let alt: u8 = named_arg(&args, "--alt")
            .and_then(|s| s.parse().ok())
            .unwrap_or(1);
        let ep_in: u8 = named_arg(&args, "--ep")
            .and_then(|s| parse_hex_u8(&s).ok())
            .unwrap_or(0x81);
        let packet_size: u32 = named_arg(&args, "--pkt-size")
            .and_then(|s| s.parse().ok())
            .unwrap_or(1024);
        let num_packets: u32 = named_arg(&args, "--num-pkts")
            .and_then(|s| s.parse().ok())
            .unwrap_or(32);

        let (vid, pid) = parse_vid_pid(vid_pid)?;
        Ok(Self {
            output,
            vid,
            pid,
            iterations,
            condition,
            iface,
            alt,
            ep_in,
            packet_size,
            num_packets,
        })
    }
}

fn default_condition() -> &'static str {
    #[cfg(not(target_family = "wasm"))]
    { "native-rusb" }
    #[cfg(all(target_family = "wasm", feature = "c4-rusb-wasi"))]
    { "wasi-rusb" }
    #[cfg(all(target_family = "wasm", not(feature = "c4-rusb-wasi")))]
    { "wasi-raw-wit" }
}

fn named_arg(args: &[String], name: &str) -> Option<String> {
    args.windows(2)
        .find(|w| w[0] == name)
        .map(|w| w[1].clone())
}

fn parse_vid_pid(s: &str) -> Result<(u16, u16)> {
    let (v, p) = s
        .split_once(':')
        .ok_or_else(|| anyhow!("expected VID:PID (hex), got {s:?}"))?;
    let vid = u16::from_str_radix(v.trim(), 16)
        .map_err(|e| anyhow!("invalid VID {v:?}: {e}"))?;
    let pid = u16::from_str_radix(p.trim(), 16)
        .map_err(|e| anyhow!("invalid PID {p:?}: {e}"))?;
    Ok((vid, pid))
}

fn parse_hex_u8(s: &str) -> Result<u8> {
    let s = s.trim_start_matches("0x").trim_start_matches("0X");
    u8::from_str_radix(s, 16).map_err(|e| anyhow!("invalid hex byte {s:?}: {e}"))
}
